proposals/0110-exponential-fee-for-write-lock-accounts.md

```diff
@@ -0,0 +1,138 @@
+
+- Account Association with Compute Unit Pricer:
+  - Accounts are associated with a *compute unit pricer*, and the *runtime*
+  maintains an LRU cache of actively contentious accounts' public keys and
+  their *compute unit pricers*.
+  - Alternatively, each account can have its *compute unit pricer* stored
+  onchain, which would require modifying accounts.
+- Compute Unit Pricer:
+  - Tracks an account's EMA *compute-unit utilization*, updated after the
+  current bank is frozen.
+  - Provides the current *cost rate* when queried.
+- EMA of Compute-Unit Utilization:
+  - Uses 150 slots for EMA calculation.
+  - EMA Alpha, representing the degree of weighting decrease, is calculated as
+  `alpha = 2 / (N+1)`.
+- Pricing Algorithm:
+  - Adjusts write-lock *cost rate* based on an account's EMA *compute-unit
+  utilization*. Initial write-lock cost rate is `1000 lamport/CU`.
+  - For each block, if an account's EMA *compute-unit utilization* is more than
+  half of its max limit, its write-lock *cost rate* increases by 1%. If it's
+  below half, the *cost rate* decreases by 1%.
+- Calculate *Write Lock Fee*:
+  - Fee required to write-lock an account is calculated by multiplying the
+  write-lock *cost rate* by the transaction's requested CU.
+
+## Detailed Design
+
+- Initialization and Inheritance:
+  - Bank initializes an empty account_write_lock_fee_cache, an LRU Cache of
+  {account_pubkey, compute-unit-pricer}.
+  - Child banks inherit the parent's cache.
+- Transaction Fee Calculation:
+  - Calculate write-lock fee for each account a transaction needs to write,
+  summing up to be its *write lock fee*. This, along with signature fee and
+  priority fee, constitutes the total fee for the transaction.
+  - Leader checks fee payer's balance before scheduling the transaction.
+- Cost Tracking:
+  - Cost_tracker tracks CUs for the current block and each write-locked accounti
+  as-is;
+  - Ensuring cost tracking is enabled at the replay stage.
+- End of Block Processing:
+  - Identify write-locked accounts with *compute-unit utilization* > half of
+  account max CU limit. Add/update bank's account_write_lock_fee_cache. 
+  - Adding new account into LRU cache could push out eldest account;
+  - LRU cache has capacity of 1024, which should be large enough for hot accounts
+  in 150 slots.
+- Fee Handling:
+  - Collected write-lock fees are 100% burnt.
+  - Collected priority fees are 100% rewarded.
+  - Collected signature fees are 50% burnt, 50% rewarded.
+
+### Other Considerations
+
+- Users may need new instruction to set a maximum write-lock fee for transaction
+- Consider tooling for wallets/simulators to query "min/max write lock fee."
+- Acknowledge read lock contention, deferring EMA fee implementation for read locks.
+- In the future, a percentage of collected write-lock-fee could be deposited
+to an account, allowing dApps to refund cranks and other service providers.
+This decision should be done via a governance vote.
+
+## Impact
```

Comment:
one addition: if fee is paid per CU requested, this additionally incentivizes accurate CU estimation beyond what the prio fee already does

Reply:
another addition (less obvious if good or bad): this increases the incentive for app developers to fit more of app state into a single account rather than having a bunch of accounts

and incentivizes some type of account-sybiling, e.g. the "canonical account for some app state" rotates every N slots to allow the fee to cool down

Reply:
Won't it have the opposite effect? If that account is saturated then fees increase super-linearly.

Reply:
(assuming this is referring to the first note in the second comment) Today a bunch of programs require always hitting accounts A B and C so the user would have to pay for 3 account writes whose price escalate together. So there's an incentive for the developer to compress the state to a single account A to decrease write-lock fee by 2/3

Reply:
@taozhu-chicago might make sense to scale it by bytes. So there is little advantage to combining accounts. 

So each account starts at a lamports per CU per byte rate (LCBR) that scales 1.01x each block. 

One way to fit this into existing fee model is to
* lower the existing signature fee by 50%
* have validators earn 100% of the signature fee 
* set the floor LCBR to match what votes currently use, so whatever vote writes to be 50% of existing signature fee. Burn 100%

So basically math works out to be the same for votes.

Reply:
account bytes are converted into CU by cost model, feel it's simpler to stay with "lamports_per_cu". As for vote, it has const CU cost, it can also have const fee, assume it is "simple vote transaction", and all "complex vote transactions" are dropped.

Reply:
complex votes can pay the complex fee :)